plugins/auth-node/src/sign-in/commonSignInResolvers.ts

```diff
@@ -40,6 +40,7 @@ export namespace commonSignInResolvers {
 
           return ctx.signInWithCatalogUser({
```

Comment:
Alright, we spoke among ourselves internally and thought that while this addresses the problem locally, there's an even deeper fix that should be made instead to benefit all code paths. I hope you feel you'd be up for giving it a go:

https://github.com/backstage/backstage/blob/b130eb36b4a00414f8763e4da7c078064850fc2d/plugins/auth-backend/src/lib/resolvers/CatalogAuthResolverContext.ts#L127

If you check this code, all of the code paths insert `kind: 'User'` automatically, _except_ the `filter` case. This is [actually documented behaviour](https://github.com/backstage/backstage/blob/ce0adea1d6477b2678aff471b4be2874bb1923e5/plugins/auth-node/src/types.ts#L89)! But we agree that it'd be an improvement to change that and instead start enforcing the kind there too.

The problem is, the `EntityFilterQuery` is a [mildly complex type](https://github.com/backstage/backstage/blob/ce0adea1d6477b2678aff471b4be2874bb1923e5/packages/catalog-client/src/types/api.ts#L74).

You could need to

- deep clone the filter
- iterate through the filter(s) (it can be an object or an array of objects)
- check if they already have a `kind` key
- if they do, move on (just assume that the caller knows what they are doing)
- if they don't, add one

so something like (untested pseudocode)

```ts
const clone = lodash.cloneDeep(filter);
for (const item of [clone].flat()) {
  if (!('kind' in item)) {
    item.kind = 'User';
  }
}
```

... or so.

Reply:
I totally agree and that was the kind of fix I wanted to go for in the beginning, but was afraid of potential side effects (the way my fix works now it's somewhat strange to have a "findCatalogUser" function and add the kind from outside to actually select the default User kind). I will take a look and make it behave more like the way it behaves when using an entityRef!

Reply:
@freben I just had a look at it and have two questions (the first one being because I'm not really used to javascript and want to understand):

1. why do we need to deep clone the filter? is it because javascript uses call by reference on complex types and we want to avoid manipulating the original objects referenced by the original filter (and if yes, does the array returned by the flat function potentially still hold references to the original objects)?
2. the EntityFilterQuery type remarks 'All matching of keys and values is case insensitive', so I assume the `('kind' in item)` might fail in situations where someone uses anything but the lowercased key, is that correct?

thank you and have a nice weekend!

Reply:
1: Correct on both counts, so as a courtesy we try to defensively try to not ruin anything for the caller.


2: Good call! I won't hold it against you if you choose the lowercase variant and call it a day for now, though there's bonus karma points for going the extra mile.